File: tokens.kt
import androidx.compose.ui.unit.sp
import androidx.compose.ui.unit.em
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.DpOffset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Shadow
import androidx.compose.ui.text.font.FontWeight

// Design Tokens - Kotlin Constants
object DesignTokens {

    // Colors


    // LightColors
    val _colors_primary_Light = Color(0xFF4F46E5)
    val _colors_primary_pressedLight = Color(0xFF4338CA)
    val _colors_primary_activeLight = Color(0xFF3730A3)
    val _colors_primary_disabledLight = Color(0xFFC7D2FE)
    val _colors_background_Light = Color(0xFFFFFFFF)
    val _colors_background_surfaceLight = Color(0xFFF9FAFB)
    val _colors_text_primaryLight = Color(0xFF111827)
    val _colors_text_secondaryLight = Color(0xFF6B7280)
    val _colors_text_disabledLight = Color(0xFF9CA3AF)
    val _colors_status_errorLight = Color(0xFFDC2626)
    val _colors_status_successLight = Color(0xFF16A34A)
    val _colors_status_warningLight = Color(0xFFD97706)
    val _colors_status_infoLight = Color(0xFF0EA5E9)

    // Typography
    // Font Families
    val light_typography_fontFamilies_main = "Inter"
    val light_typography_fontFamilies_display = "Poppins"

    // Font Weights
    val light_typography_fontWeights_regular = FontWeight.Normal
    val light_typography_fontWeights_medium = FontWeight.Medium
    val light_typography_fontWeights_bold = FontWeight.Bold

    // Font Sizes
    val light_typography_fontSizes_sm = 14.sp
    val light_typography_fontSizes_md = 16.sp
    val light_typography_fontSizes_lg = 20.sp
    val light_typography_fontSizes_xl = 24.sp

    // Line Heights
    val light_typography_lineHeights_tight = 1.25f
    val light_typography_lineHeights_normal = 1.5f
    val light_typography_lineHeights_relaxed = 1.625f

    // Spacing
    val light_spacing_sm = 8.dp
    val light_spacing_md = 16.dp
    val light_spacing_lg = 24.dp
    val light_spacing_xl = 32.dp

    // Sizing
    val light_sizing_buttonHeight = 44.dp
    val light_sizing_inputHeight = 48.dp
    val light_sizing_cardHeight = 120.dp
    val light_sizing_sidebarWidth = 280.dp
    val light_sizing_modalWidth = 400.dp

    // Border Radius
    val light_borderRadius_sm = 4.dp
    val light_borderRadius_md = 8.dp
    val light_borderRadius_lg = 16.dp
    val light_borderRadius_full = 9999.dp

    // Borders
    val light_border_thin = 1.dp
    val light_border_medium = 2.dp
    val light_border_thick = 2.dp

    // Shadows
    // Small Shadows
    val light_shadows_sm = Shadow(
        color = Color(0px).copy(alpha = 0.25f),
        offset = DpOffset(0.dp, 1.dp),
        blurRadius = 3.dp
    )

    val light_shadows_md = Shadow(
        color = Color(-1px).copy(alpha = 0.25f),
        offset = DpOffset(0.dp, 4.dp),
        blurRadius = 6.dp
    )


    // Z-Indices
    val light_zIndex_dropdown = 1000f
    val light_zIndex_modal = 1050f
    val light_zIndex_tooltip = 1100f

    // Transitions
    val light_transitions_fast = 150L
    val light_transitions_normal = 200L
    val light_transitions_slow = 300L

    // Other Tokens
    val light_icons_sm = "16"
    val light_icons_md = "20"
    val light_icons_lg = "24"

}

File: tokens.swift
// Design Tokens - SwiftUI with Structs and Enums
import SwiftUI

// MARK: - Colors
enum Colors {
    static let colorsPrimaryLight = Color(hex: "#4F46E5")
    static let colorsPrimaryPressedLight = Color(hex: "#4338CA")
    static let colorsPrimaryActiveLight = Color(hex: "#3730A3")
    static let colorsPrimaryDisabledLight = Color(hex: "#C7D2FE")
    static let colorsBackgroundLight = Color(hex: "#FFFFFF")
    static let colorsBackgroundSurfaceLight = Color(hex: "#F9FAFB")
    static let colorsTextPrimaryLight = Color(hex: "#111827")
    static let colorsTextSecondaryLight = Color(hex: "#6B7280")
    static let colorsTextDisabledLight = Color(hex: "#9CA3AF")
    static let colorsStatusErrorLight = Color(hex: "#DC2626")
    static let colorsStatusSuccessLight = Color(hex: "#16A34A")
    static let colorsStatusWarningLight = Color(hex: "#D97706")
    static let colorsStatusInfoLight = Color(hex: "#0EA5E9")
}

// MARK: - Typography
enum Typography {
    enum FontFamily {
        static let lightTypographyFontFamiliesMain = "Inter"
        static let lightTypographyFontFamiliesDisplay = "Poppins"
    }

    enum FontSize {
        static let lightTypographyFontSizesSm: CGFloat = 14
        static let lightTypographyFontSizesMd: CGFloat = 16
        static let lightTypographyFontSizesLg: CGFloat = 20
        static let lightTypographyFontSizesXl: CGFloat = 24
    }

    enum FontWeight {
        static let lightTypographyFontWeightsRegular: Int = 400
        static let lightTypographyFontWeightsMedium: Int = 500
        static let lightTypographyFontWeightsBold: Int = 700
    }

    enum LineHeight {
        static let lightTypographyLineHeightsTight: CGFloat = 1.25
        static let lightTypographyLineHeightsNormal: CGFloat = 1.5
        static let lightTypographyLineHeightsRelaxed: CGFloat = 1.625
    }

}

// MARK: - Spacing
enum Spacing {
    static let lightSpacingSm: CGFloat = 8
    static let lightSpacingMd: CGFloat = 16
    static let lightSpacingLg: CGFloat = 24
    static let lightSpacingXl: CGFloat = 32
}

// MARK: - Sizing
enum Sizing {
    static let lightSizingButtonHeight: CGFloat = 44
    static let lightSizingInputHeight: CGFloat = 48
    static let lightSizingCardHeight: CGFloat = 120
    static let lightSizingSidebarWidth: CGFloat = 280
    static let lightSizingModalWidth: CGFloat = 400
}

// MARK: - Border Radius
enum BorderRadius {
    static let lightBorderRadiusSm: CGFloat = 4
    static let lightBorderRadiusMd: CGFloat = 8
    static let lightBorderRadiusLg: CGFloat = 16
    static let lightBorderRadiusFull: CGFloat = 9999
}

// MARK: - Shadows
enum Shadows {
    static let lightShadowsSm = ShadowStyle(
        color: Color(hex: "0px").opacity(Opacities.shadow),
        radius: 3,
        x: 0,
        y: 1
    )

    static let lightShadowsMd = ShadowStyle(
        color: Color(hex: "-1px").opacity(Opacities.shadow),
        radius: 6,
        x: 0,
        y: 4
    )

}

// MARK: - Z-Indices
enum ZIndices {
    static let lightZIndexDropdown: Double = 1000
    static let lightZIndexModal: Double = 1050
    static let lightZIndexTooltip: Double = 1100
}

// MARK: - Transitions
enum Transitions {
    static let fast: Double = 0.15
    static let medium: Double = 0.2
    static let slow: Double = 0.3
    // Durations (in seconds)
    static let lightTransitionsFast: Double = 0.15
    static let lightTransitionsNormal: Double = 0.2
    static let lightTransitionsSlow: Double = 0.3

}

// MARK: - Other Tokens
enum OtherTokens {
    static let lightBorderThin = "1"
    static let lightBorderMedium = "1.5"
    static let lightBorderThick = "2"
    static let lightIconsSm = "16"
    static let lightIconsMd = "20"
    static let lightIconsLg = "24"
}

// MARK: - Color Extensions
extension Color {
    init(hex: String) {
        let hex = hex.trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
        var int: UInt64 = 0
        Scanner(string: hex).scanHexInt64(&int)
        let a, r, g, b: UInt64
        switch hex.count {
        case 3: // RGB (12-bit)
            (a, r, g, b) = (255, (int >> 8) * 17, (int >> 4 & 0xF) * 17, (int & 0xF) * 17)
        case 6: // RGB (24-bit)
            (a, r, g, b) = (255, int >> 16, int >> 8 & 0xFF, int & 0xFF)
        case 8: // ARGB (32-bit)
            (a, r, g, b) = (int >> 24, int >> 16 & 0xFF, int >> 8 & 0xFF, int & 0xFF)
        default:
            (a, r, g, b) = (1, 1, 1, 0)
        }

        self.init(
            .sRGB,
            red: Double(r) / 255,
            green: Double(g) / 255,
            blue: Double(b) / 255,
            opacity: Double(a) / 255
        )
    }

    init(light: String, dark: String) {
        self.init(UIColor { traitCollection in
            switch traitCollection.userInterfaceStyle {
            case .dark:
                return UIColor(Color(hex: dark))
            default:
                return UIColor(Color(hex: light))
            }
        })
    }
}

// MARK: - ShadowStyle
struct ShadowStyle {
    let color: Color
    let radius: CGFloat
    let x: CGFloat
    let y: CGFloat
    
    init(color: Color, radius: CGFloat, x: CGFloat = 0, y: CGFloat = 0) {
        self.color = color
        self.radius = radius
        self.x = x
        self.y = y
    }
}

// MARK: - View Extension for Shadows
extension View {
    func shadow(_ shadowStyle: ShadowStyle) -> some View {
        self.shadow(
            color: shadowStyle.color,
            radius: shadowStyle.radius,
            x: shadowStyle.x,
            y: shadowStyle.y
        )
    }
}

// MARK: - Line Height Modifier
struct LineHeightModifier: ViewModifier {
    let fontSize: CGFloat
    let lineHeight: CGFloat
    func body(content: Content) -> some View {
        let target = lineHeight * fontSize
        let spacing = max(0, target - fontSize)
        return content
            .lineSpacing(spacing)
            .padding(.vertical, spacing / 2)
    }
}

extension View {
    func lineHeight(_ lineHeight: CGFloat, fontSize: CGFloat) -> some View {
        modifier(LineHeightModifier(fontSize: fontSize, lineHeight: lineHeight))
    }
}

// MARK: - Letter Spacing (em) Modifier
struct LetterSpacingEmModifier: ViewModifier {
    let fontSize: CGFloat
    let letterSpacingEm: CGFloat
    func body(content: Content) -> some View {
        let trackingPoints = letterSpacingEm * fontSize
        return content.tracking(trackingPoints)
    }
}

extension View {
    func letterSpacingEm(_ letterSpacingEm: CGFloat, fontSize: CGFloat) -> some View {
        modifier(LetterSpacingEmModifier(fontSize: fontSize, letterSpacingEm: letterSpacingEm))
    }
}

// MARK: - Font Weight Mapping (Int -> Font.Weight)
extension Font.Weight {
    static func fromNumeric(_ value: Int) -> Font.Weight {
        switch value {
        case ..<150: return .ultraLight
        case 150..<250: return .thin
        case 250..<350: return .light
        case 350..<450: return .regular
        case 450..<550: return .medium
        case 550..<650: return .semibold
        case 650..<750: return .bold
        case 750..<850: return .heavy
        default: return .black
        }
    }
}

extension View {
    func fontWeight(_ numericWeight: Int) -> some View {
        self.fontWeight(Font.Weight.fromNumeric(numericWeight))
    }
}

// MARK: - Animation Extensions
extension Animation {
    static let fast = Animation.easeInOut(duration: Transitions.fast)
    static let medium = Animation.easeInOut(duration: Transitions.medium)
    static let slow = Animation.easeInOut(duration: Transitions.slow)
}

